go/store/nbs/table_index.go

```diff
@@ -479,188 +556,138 @@ func (ti onHeapTableIndex) Clone() (tableIndex, error) {
-	for idx := i.prefixIdx(prefix); idx < i.chunkCount && i.prefixes[idx] == prefix; idx++ {
-		mi := idx * mmapIndexEntrySize
-		e := mmapIndexEntry(i.data[mi : mi+mmapIndexEntrySize])
-		if bytes.Equal(e.suffix(), h[addrPrefixSize:]) {
-			return e, true, nil
+// Close closes the underlying onHeapTableIndex and then unmaps the memory
+// region.
+func (ti *mmapTableIndex) Close() error {
+	cnt := atomic.AddInt32(ti.refCnt, -1)
+	if cnt == 0 {
+		chunkCount := ti.chunkCount
+		// mmapTableIndex sets the quota provider for onHeapTableIndex to a
+		// noopQuotaProvider, so that we can release quota after the memory region
+		// is unmapped.
+		err := ti.onHeapTableIndex.Close()
+		if err != nil {
+			return err
 		}
-	}
-	return mmapIndexEntry{}, false, nil
-}
 
-func (i mmapTableIndex) Ordinals() ([]uint32, error) {
-	s := mmapOrdinalSlice(make([]mmapOrdinal, i.chunkCount))
-	for idx := 0; uint32(idx) < i.chunkCount; idx++ {
-		mi := idx * mmapIndexEntrySize
-		e := mmapIndexEntry(i.data[mi : mi+mmapIndexEntrySize])
-		s[idx] = mmapOrdinal{idx, e.Offset()}
+		ti.indexDataBuff = nil
+		ti.offset1DataBuff = nil
+		err = ti.mmapped.Unmap()
+		if err != nil {
+			return err
+		}
+
+		err = ti.q.ReleaseQuota(indexMemSize(chunkCount))
+		if err != nil {
+			return err
+		}
 	}
-	sort.Sort(s)
-	res := make([]uint32, i.chunkCount)
-	for j, r := range s {
-		res[r.idx] = uint32(j)
+	if cnt < 0 {
+		panic("Close() called and reduced ref count to < 0.")
 	}
-	return res, nil
+
+	return nil
 }
 
-func (i mmapTableIndex) Prefixes() ([]uint64, error) {
-	return i.prefixes, nil
+func (ti *mmapTableIndex) parseIndexBuffer(q MemoryQuotaProvider) (err error) {
+	ti.onHeapTableIndex, err = parseTableIndexWithOffsetBuff(ti.indexDataBuff, ti.offset1DataBuff, &noopQuotaProvider{})
+	ti.q = q
+	return err
 }
 
-func (i mmapTableIndex) TableFileSize() uint64 {
-	return i.fileSz
+type notifyFunc func(n uint64, total uint64)
+
+var noOpNotify = func(uint64, uint64) {}
+
+type mmapStats struct {
+	mu        sync.Mutex
+	totalUsed uint64
+	WillMmap  notifyFunc
+	Mmapped   notifyFunc
+	UnMapped  notifyFunc
 }
 
-func (i mmapTableIndex) TotalUncompressedData() uint64 {
-	return i.totalUncompressedData
+var GlobalMmapStats = &mmapStats{
+	sync.Mutex{},
+	0,
+	noOpNotify,
+	noOpNotify,
+	noOpNotify,
 }
 
-func (i mmapTableIndex) Close() error {
-	cnt := atomic.AddInt32(i.refCnt, -1)
-	if cnt == 0 {
-		return i.data.Unmap()
-	}
-	if cnt < 0 {
-		panic("Close() called and reduced ref count to < 0.")
-	}
-	return nil
+type mmapWStat struct {
+	m    mmap.MMap
+	used uint64
 }
 
-func (i mmapTableIndex) Clone() (tableIndex, error) {
-	cnt := atomic.AddInt32(i.refCnt, 1)
-	if cnt == 1 {
-		panic("Clone() called after last Close(). This index is no longer valid.")
+func mmapWithStats(f *os.File, length int, prot, flags int, offset int64) (mmapWStat, error) {
+	GlobalMmapStats.mu.Lock()
+	GlobalMmapStats.WillMmap(uint64(length), GlobalMmapStats.totalUsed)
```

Comment:
`defer GlobalMmapStats.mu.Unlock()` (or Unlock it in the error early return for MapRegion)